File: client/Ciphelet/app/src/main/java/com/github/worldoneo/ciphelet/connector/ActionClient.java
package com.github.worldoneo.ciphelet.connector;

import androidx.core.util.Consumer;

import com.github.worldoneo.ciphelet.connector.action.GenericAction;

import org.java_websocket.client.WebSocketClient;
import org.java_websocket.drafts.Draft;
import org.java_websocket.handshake.ServerHandshake;

import java.net.URI;
import java.util.Map;

public class ActionClient extends WebSocketClient {

    private Consumer<GenericAction> actionConsumer;

    public ActionClient(URI serverUri, Draft draft) {
        super(serverUri, draft);
    }

    public ActionClient(URI serverURI) {
        super(serverURI);
    }

    public ActionClient(URI serverUri, Map<String, String> httpHeaders) {
        super(serverUri, httpHeaders);
    }

    @Override
    public void onOpen(ServerHandshake handshakedata) {
    }

    @Override
    public void onMessage(String message) {
        System.out.println("received: " + message);
        if (actionConsumer != null) {
            GenericAction genericAction = Connector.gson.fromJson(message, GenericAction.class);
            actionConsumer.accept(genericAction);
        }
    }

    @Override
    public void onClose(int code, String reason, boolean remote) {
    }

    @Override
    public void onError(Exception ex) {
        ex.printStackTrace();
        // if the error is fatal then onClose will be called additionally
    }

    public void setActionConsumer(Consumer<GenericAction> actionConsumer) {
        this.actionConsumer = actionConsumer;
    }
}

File: client/Ciphelet/app/src/main/java/com/github/worldoneo/ciphelet/connector/ActionClient.kt
package com.github.worldoneo.ciphelet.connector

import androidx.core.util.Consumer
import com.github.worldoneo.ciphelet.connector.action.GenericAction
import org.java_websocket.client.WebSocketClient
import org.java_websocket.handshake.ServerHandshake
import java.net.URI

class ActionClient(serverURI: URI?) : WebSocketClient(serverURI) {
    private var actionConsumer: Consumer<GenericAction>? = null

    override fun onOpen(handshakedata: ServerHandshake) {}
    override fun onMessage(message: String) {
        println("received: $message")
        if (actionConsumer != null) {
            val genericAction = Connector.gson.fromJson(message, GenericAction::class.java)
            actionConsumer!!.accept(genericAction)
        }
    }

    override fun onClose(code: Int, reason: String, remote: Boolean) {}
    override fun onError(ex: Exception) {
        ex.printStackTrace()
        // if the error is fatal then onClose will be called additionally
    }

    fun setActionConsumer(actionConsumer: Consumer<GenericAction>) {
        this.actionConsumer = actionConsumer
    }
}